### PR TITLE
[WASMFS] Switch to using `out` and `err`

### PR DESCRIPTION
**Pull request by ethanalee**
- Propose using `out` and `err` instead of `emscripten_console_log` and `emscripten_console_error`

Similar issue to #14805, where pthreads + node led to inconsistent output with `emscripten_console_log`.

Add `emscripten_out` and `emscripten_err` along with `f` versions.
Replaced `emscripten_console_log`/`error` with those two functions respectively in `streams.cpp`.

### CODE REVIEW

File: system/include/emscripten/console.h [hunk 1]
```diff
@@ -20,6 +20,12 @@ void emscripten_console_logf(const char *utf8String, ...) __attribute__((__forma
 void emscripten_console_warnf(const char *utf8String, ...) __attribute__((__format__(printf, 1, 2)));
 void emscripten_console_errorf(const char *utf8String, ...)__attribute__((__format__(printf, 1, 2)));
 
+// Write to the out() and err() hooks directly.
+// out() and err() are defined in shell.js.
```
Comment (sbc100) [inline system/include/emscripten/console.h:24]: Perhaps say something like.  "out() and err() are defined in shell.js and have different behaviour from console.log/error.  For example under node they write the stderr and stdout streams which, especially in workers, are more direct ways to write output".

Comment (sbc100) [reply system/include/emscripten/console.h:24]: I kind of wish we had better names for these ... but I can't think of any.

**Review by sbc100 — APPROVED** [review]
lgtm with comments/rationale added.

File: system/include/emscripten/console.h (hunk 1 above)
Comment (ethanalee) [reply system/include/emscripten/console.h]: Done.

File: system/lib/wasmfs/streams.cpp [hunk 2]
```diff
@@ -34,7 +34,7 @@ std::shared_ptr<StdinFile> StdinFile::getSingleton() {
 }
 
 __wasi_errno_t StdoutFile::write(const uint8_t* buf, size_t len, off_t offset) {
-  return writeStdBuffer(buf, len, &emscripten_console_log, writeBuffer);
+  return writeStdBuffer(buf, len, &emscripten_out, writeBuffer);
```
Comment (kripken) [inline system/lib/wasmfs/streams.cpp:37]: please add a link to the issue with node that is the reason for using out/err here.

File: system/include/emscripten/console.h [hunk 3]
```diff
@@ -20,6 +20,14 @@ void emscripten_console_logf(const char *utf8String, ...) __attribute__((__forma
 void emscripten_console_warnf(const char *utf8String, ...) __attribute__((__format__(printf, 1, 2)));
 void emscripten_console_errorf(const char *utf8String, ...)__attribute__((__format__(printf, 1, 2)));
 
+// Write to the out() and err() hooks directly (defined in shell.js).
+// These have different behaviour compared to console.log/err.
+// Under node, they write to stdout and stderr which is a more direct way to
+// write output especially in worker threads.
+// See https://github.com/emscripten-core/emscripten/issues/14804
```
Comment (kripken) [inline system/include/emscripten/console.h:27]: They are also aware of `Module.print / Module.printErr`, which is worth mentioning too. Given that I would basically describe them as writing to the standard out or standard error streams, including user overrides.

**Review by kripken — COMMENTED** [review]
I am a little unsure about adding a new public API here, but otherwise this lgtm.

File: system/include/emscripten/console.h (hunk 3 above)
Comment (sbc100) [reply system/include/emscripten/console.h:27]: Yes, it I agree its worth mentioning the module overrides.   How about adding one more sentence such as:  "The default behavior of these functions can be overridden by `print` and `printErr`, if provided on the Module object."

**Comment by sbc100** [comment]
> I am a little unsure about adding a new public API here, but otherwise this lgtm.

I agree we could keep these as internal for now... but also the err() and out() methods are very widespread in the library code and it makes sense to make them accessible from native code too I think.

Also, perhaps you could add the `f` versions for convenience.   I would find those useful for debugging I think.

**Comment by sbc100** [comment]
> > I am a little unsure about adding a new public API here, but otherwise this lgtm.
> 
> I agree we could keep these as internal for now... but also the err() and out() methods are very widespread in the library code and it makes sense to make them accessible from native code too I think.
> 
> Also, perhaps you could add the `f` versions for convenience. I would find those useful for debugging I think.

On balance, I think I would at least like these functions to be available internally to all system libraryies, and I would also like to be able to use them in tests.   We don't really have internal-only headers today, but we could invent such a concept.   But these are widely useful internally its likely they are also useful externally so I think there is a good argument for just making them public..

**Comment by kripken** [comment]
@sbc100 Sounds reasonable. I guess we could prefix them with `_` if we wanted to hint that they are internal. But yeah, as you said there are reasonable arguments the other way. Sgtm as everyone else prefers.

**Comment by ethanalee** [comment]
Added the `_` prefix and added the `f` version as well.

**Review by sbc100 — APPROVED** [review]

**Review by tlively — APPROVED** [review]

File: src/library.js [hunk 4]
```diff
@@ -3494,6 +3494,22 @@ LibraryManager.library = {
 #endif
   },
 
+  emscripten_out__sig: 'vi',
+  emscripten_out: function(str) {
```
Comment (tlively) [inline src/library.js:3498]: Do these need the `_` prefix to match the C++ names?
```suggestion
  _emscripten_out__sig: 'vi',
  _emscripten_out: function(str) {
```

**Pull request merged**